### ISSUE TITLE
Callbacks are not being called, gradient norm not getting logged,how to debug these scenarios? 

### ISSUE THREAD

**Issue by dchatterjee172** [PyTorchLightning/pytorch-lightning#3065]
## ❓ Questions and Help
* I have defined a callback for the `on_epoch_end` event. Also, I have defined `track_grad_norm = 1`, which I think is also kind of a callback.
* When I used `--fast_dev_run` or `--limit_train_batches` and `--limit_val_batches` the callback(including gradient norm plot) got executed properly.
* When I run without the above flags, `on_epoch_end` is sometimes not executing (or silently failing?), gradient norms are not getting plotted at all. 
* Checkpoint saving callback is always successfully executing.

Now I am unsure how do I debug this scenario. There is no error messages logged.
#### What's your environment?

 - OS: Linux
 - Packaging:- pip
 - Version:- 0.9.0rc15
